### ISSUE TITLE
Update-PSResource tries to reinstall required dependencies

### ISSUE THREAD

**Issue by ThomasNieto** [PowerShell/PSResourceGet#1273]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://www.powershellgallery.com/packages/PSResourceGet)
- [X] Search the existing issues.

### Steps to reproduce

Update a module with a required dependency, it is helpful to have a binary module loaded so the files are locked.

### Expected behavior

```console
# required module is not installed again
```

### Actual behavior

```console
Update-PSResource anypackage.programs -PassThru
Update-PSResource: Unable to successfully install package 'AnyPackage': 'Cannot remove package path C:\Users\Thomas\Documents\PowerShell\Modules\AnyPackage\0.5.1. The previous package contents have been restored.'
Update-PSResource: Operation is not valid due to the current state of the object.
```

### Error details

```console
ErrorIndex: 0

Exception             :
    Type    : System.InvalidOperationException
    Message : Operation is not valid due to the current state of the object.
    HResult : -2146233079
TargetObject          : Microsoft.PowerShell.PSResourceGet.Cmdlets.InstallHelper
CategoryInfo          : InvalidOperation: (Microsoft.PowerShel…dlets.InstallHelper:InstallHelper) [Update-PSResource],
InvalidOperationException
FullyQualifiedErrorId : InstallPackageTryMoveContentFailure,Microsoft.PowerShell.PSResourceGet.Cmdlets.UpdatePSResource
InvocationInfo        :
    MyCommand        : Update-PSResource
    ScriptLineNumber : 1
    OffsetInLine     : 1
    HistoryId        : 1
    Line             : Update-PSResource anypackage.programs -PassThru
    Statement        : Update-PSResource anypackage.programs -PassThru
    PositionMessage  : At line:1 char:1
                       + Update-PSResource anypackage.programs -PassThru
                       + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    InvocationName   : Update-PSResource
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo :




   ErrorIndex: 1

Exception             :
    Type           : System.Management.Automation.PSInvalidOperationException
    ErrorRecord    :
        Exception             :
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : Unable to successfully install package 'AnyPackage': 'Cannot remove package path
C:\Users\Thomas\Documents\PowerShell\Modules\AnyPackage\0.5.1. The previous package contents have been restored.'
            HResult : -2146233087
        CategoryInfo          : InvalidOperation: (:) [], ParentContainsErrorRecordException
        FullyQualifiedErrorId : InvalidOperation
    Message        : Unable to successfully install package 'AnyPackage': 'Cannot remove package path
C:\Users\Thomas\Documents\PowerShell\Modules\AnyPackage\0.5.1. The previous package contents have been restored.'
    InnerException :
        Type           : System.Management.Automation.PSInvalidOperationException
        ErrorRecord    :
            Exception             :
                Type    : System.Management.Automation.ParentContainsErrorRecordException
                Message : Cannot remove package path C:\Users\Thomas\Documents\PowerShell\Modules\AnyPackage\0.5.1.
The previous package contents have been restored.
                HResult : -2146233087
            CategoryInfo          : InvalidOperation: (:) [], ParentContainsErrorRecordException
            FullyQualifiedErrorId : InvalidOperation
        TargetSite     :
            Name          : DeleteDirectoryWithRestore
            DeclaringType : Microsoft.PowerShell.PSResourceGet.UtilClasses.Utils, Microsoft.PowerShell.PSResourceGet,
Version=0.5.22.0, Culture=neutral, PublicKeyToken=null
            MemberType    : Method
            Module        : Microsoft.PowerShell.PSResourceGet.dll
        Message        : Cannot remove package path C:\Users\Thomas\Documents\PowerShell\Modules\AnyPackage\0.5.1. The
previous package contents have been restored.
        InnerException :
            Type       : System.UnauthorizedAccessException
            TargetSite :
                Name          : DeleteFile
                DeclaringType : System.IO.FileSystem, System.Private.CoreLib, Version=8.0.0.0, Culture=neutral,
PublicKeyToken=7cec85d7bea7798e
                MemberType    : Method
                Module        : System.Private.CoreLib.dll
            Message    : Access to the path
'C:\Users\Thomas\Documents\PowerShell\Modules\AnyPackage\0.5.1\AnyPackage.dll' is denied.
            Source     : System.Private.CoreLib
            HResult    : -2147024891
            StackTrace :
   at System.IO.FileSystem.DeleteFile(String fullPath)
   at Microsoft.PowerShell.PSResourceGet.UtilClasses.Utils.DeleteDirectory(String dirPath)
   at Microsoft.PowerShell.PSResourceGet.UtilClasses.Utils.DeleteDirectoryWithRestore(String dirPath)
        Source         : Microsoft.PowerShell.PSResourceGet
        HResult        : -2146233079
        StackTrace     :
   at Microsoft.PowerShell.PSResourceGet.UtilClasses.Utils.DeleteDirectoryWithRestore(String dirPath)
   at Microsoft.PowerShell.PSResourceGet.Cmdlets.InstallHelper.MoveFilesIntoInstallPath(PSResourceInfo pkgInfo,
Boolean isModule, Boolean isLocalRepo, String dirNameVersion, String tempInstallPath, String installPath, String
newVersion, String moduleManifestVersion, String scriptPath)
   at Microsoft.PowerShell.PSResourceGet.Cmdlets.InstallHelper.TryMoveInstallContent(String tempInstallPath, ScopeType
scope, Hashtable packagesHash)
    HResult        : -2146233079
TargetObject          : Microsoft.PowerShell.PSResourceGet.Cmdlets.UpdatePSResource
CategoryInfo          : InvalidOperation: (Microsoft.PowerShel…ts.UpdatePSResource:UpdatePSResource)
[Update-PSResource], PSInvalidOperationException
FullyQualifiedErrorId : InstallPackageFailed,Microsoft.PowerShell.PSResourceGet.Cmdlets.UpdatePSResource
InvocationInfo        :
    MyCommand        : Update-PSResource
    ScriptLineNumber : 1
    OffsetInLine     : 1
    HistoryId        : 1
    Line             : Update-PSResource anypackage.programs -PassThru
    Statement        : Update-PSResource anypackage.programs -PassThru
    PositionMessage  : At line:1 char:1
                       + Update-PSResource anypackage.programs -PassThru
                       + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    InvocationName   : Update-PSResource
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo :
```

### Environment data

```powershell
beta22
```

**Comment by SydneyhSmith** [comment]
Thanks @ThomasNieto, based on your error messages, I am assuming "AnyPackage" is a dependency of "AnyPackage.Programs", was the version required of "AnyPackage" incremented in a way that would require it to be reinstalled, or was it attempted to be reinstalled unnecessarily?

**Comment by ThomasNieto** [comment]
@SydneyhSmith I'm updating `AnyPackage.Programs` which requires `AnyPackage` and that module is already installed and meets version requirements and there is not a newer version.

From the error message you can see its trying to remove `AnyPackage` module file paths which there is no need to do so since its already installed.

```
Unable to successfully install package 'AnyPackage': 'Cannot remove package path
C:\Users\Thomas\Documents\PowerShell\Modules\AnyPackage\0.5.1. The previous package contents have been restored.'
```

**Comment by ThomasNieto** [comment]
@SydneyhSmith here is another example with more detailed reproduction.

```powershell
PS C:\> Find-PSResource anypackage

Name       Version Prerelease Repository Description
----       ------- ---------- ---------- -----------
AnyPackage 0.5.1              PSGallery  Use various package managers with a single set of commands.

PS C:\> Get-InstalledPSResource anypackage

Name       Version Prerelease Repository Description
----       ------- ---------- ---------- -----------
AnyPackage 0.5.1              PSGallery  Use various package managers with a single set of commands.
AnyPackage 0.4.1              PSGallery  Use various package managers with a single set of commands.
AnyPackage 0.4.0              PSGallery  Use various package managers with a single set of commands.
AnyPackage 0.3.0              PSGallery  Use various package managers with a single set of commands.
AnyPackage 0.2.0              PSGallery  Use various package managers with a single set of commands.
AnyPackage 0.1.2              PSGallery  Use various package managers with a single set of commands.
AnyPackage 0.1.0              PSGallery  Use various package managers with a single set of commands.

PS C:\> Update-PSResource AnyPackage.Msu -PassThru
Update-PSResource: Unable to successfully install package 'AnyPackage': 'Cannot remove package path C:\Users\Thomas\Documents\PowerShell\Modules\AnyPackage\0.5.1. The previous package contents have been restored.'
Update-PSResource: Operation is not valid due to the current state of the object.
PS C:\> Get-InstalledPSResource anypackage.Msu

Name           Version Prerelease Repository Description
----           ------- ---------- ---------- -----------
AnyPackage.Msu 0.2.3              PSGallery  Windows Msu provider for AnyPackage.
AnyPackage.Msu 0.2.2              PSGallery  Windows Msu provider for AnyPackage.

PS C:\> Find-PSResource anypackage.msu

Name           Version Prerelease Repository Description
----           ------- ---------- ---------- -----------
AnyPackage.Msu 0.3.0              PSGallery  Windows Msu provider for AnyPackage.

PS C:\> Update-PSResource AnyPackage.Msu -PassThru
Update-PSResource: Unable to successfully install package 'AnyPackage': 'Cannot remove package path C:\Users\Thomas\Documents\PowerShell\Modules\AnyPackage\0.5.1. The previous package contents have been restored.'
Update-PSResource: Operation is not valid due to the current state of the object.
PS C:\> Get-InstalledPSResource anypackage.Msu

Name           Version Prerelease Repository Description
----           ------- ---------- ---------- -----------
AnyPackage.Msu 0.2.3              PSGallery  Windows Msu provider for AnyPackage.
AnyPackage.Msu 0.2.2              PSGallery  Windows Msu provider for AnyPackage.

PS C:\> Update-PSResource AnyPackage.Msu -PassThru -SkipDependencyCheck

Name           Version Prerelease Repository Description
----           ------- ---------- ---------- -----------
AnyPackage.Msu 0.3.0              PSGallery  Windows Msu provider for AnyPackage.
```

**Comment by SydneyhSmith** [comment]
@ThomasNieto thanks!

**Comment by adityapatwardhan** [comment]
Could be related to: https://github.com/PowerShell/PSResourceGet/issues/1272

**Comment by anamnavi** [comment]
This reproduces with beta22, but not with our changes for beta23.

**Comment by anamnavi** [comment]
Was fixed by PR #1255 I believe
